clash-protocols/src/Protocols/BiDf.hs

```diff
@@ -0,0 +1,139 @@
+{-# OPTIONS_GHC -fplugin Protocols.Plugin #-}
+
+-- | Bi-directional request/response-style 'Df' channels.
+module Protocols.BiDf (
+  BiDf,
+    -- * Conversion
+  fromDfs,
+  toDfs,
+  fromBiDf,
+  toBiDf,
+    -- * Trivial combinators
+  void,
+  loopback,
+    -- * Mapping
+  dimap,
+  -- * Fan-in
+  fanin
+) where
+
+import Prelude ()
+
+import Clash.Prelude
+
+import Protocols
+import qualified Protocols.Df as Df
+
+-- | A 'Protocol' allowing requests to be passed downstream, with corresponding
+-- responses being passed back upstream. Responses are provided in the order that
+-- their corresponding requests were submitted.
+--
+-- *Correctness conditions*
+--
+--  - The response channel must not produce a value before the request channel
+--    has produced a value.
```

Comment:
I think it would be good to really spell out what _produced_ and _before_ mean in this sentence.

Scenario: In cycle N, the forward channel asserts `Data Request`. The recipient doesn't yet acknowledge this, but it starts to work on the response. Then in cycle N+1, it both acknowledges `Data Request` and also produces, on the backward channel, `Data Response`.

If we say the moment it is acked, the channel has produced a value (before that, the _component_ could well be said to have produced a value, but the _channel_ is really up to interpretation), we now have the situation where the response is produced at the exact same moment as the request is produced. So it wasn't _before_, but as a reader of this documentation, I am left wondering if this is the correct interpretation.

The scenario seems to be desired behaviour, so let's rephrase or give an example to really show that this is within spec.